### DISCUSSION TITLE
How to create and tetrahedralize spherical shell?

### DISCUSSION THREAD

**Discussion by liblaf** [pyvista/pyvista#4425] [Q&A] [answered]
I'm trying to merge two spheres with different radius and pass the resulting spherical shell to TetGen. However, I found that the normal and indices of the inner sphere are not inverted. TetGen adds unnecessary points and tetrahedrons inside the inner sphere. How can I remove those points and tetrahedrons?

**Comment by tkoyama010** [comment]
I reproduce what you asking. I am thinking about how to delete cells, so please wait.
```python
"""
Using external modules to create a mesh
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Tetrahedralize a sphere
"""

# sphinx_gallery_thumbnail_number = 2
import pyvista as pv
import tetgen

###############################################################################
# Using PyVista
# ~~~~~~~~~~~~~
# Create a surface mesh using ``pyvista`` and then tetrahedralize it.
sphere_a = pv.Sphere(radius=1.0)
sphere_b = pv.Sphere(radius=0.5)
sphere_b.flip_normals()
sphere = sphere_a.merge(sphere_b)
tet = tetgen.TetGen(sphere)
tet.tetrahedralize(order=1, mindihedral=20, minratio=1.5)
grid = tet.grid
grid.plot(show_edges=True)


###############################################################################
# Use pyvista to plot

# get cell centroids
cell_center = grid.cell_centers().points

# extract cells below the 0 xy plane
mask = cell_center[:, 2] < 0
cell_ind = mask.nonzero()[0]
subgrid = grid.extract_cells(cell_ind)

# advanced plotting
plotter = pv.Plotter(off_screen=True)
plotter.add_mesh(subgrid, "lightgrey", lighting=True, show_edges=True)
plotter.add_mesh(sphere, "r", "wireframe")
plotter.add_legend([[" Input Mesh ", "r"], [" Tessellated Mesh ", "black"]])
plotter.screenshot("sphere.png")
```
![sphere](https://github.com/pyvista/pyvista/assets/7513610/dbf48d52-e9e3-42dc-9a28-b362f83833de)

**Reply by liblaf** [reply, marked as answer]
I found out that `*.smesh` files contain a list of "holes" that will be removed after tetgen (see [TetGen: File Formats](https://www.wias-berlin.de/software/tetgen/fformats.smesh.html)). This works for me. I just had to manually convert the mesh to that file format and use the tetgen executable to do the job.
